libs/base/Data/List/Quantifiers.idr

```diff
@@ -187,3 +187,63 @@ pushOutInInverse [] = Refl
 pushOutInInverse (Element x p :: xs) with (pushOutInInverse xs)
   pushOutInInverse (Element x p :: xs) | subprf with (pullOut xs)
     pushOutInInverse (Element x p :: xs) | subprf | Element ss ps = rewrite subprf in Refl
+
+------------------------------------------------------------------------
+-- Partitioning lists according to All
+
+||| Two lists, `xs` and `ys`, whose elements are interleaved in the list `xys`.
+public export
+data Interleaving : (xs, ys, xys : List a) -> Type where
+  Nil   : Interleaving [] [] []
+  Left  : Interleaving xs ys xys -> Interleaving (x :: xs) ys (x :: xys)
+  Right : Interleaving xs ys xys -> Interleaving xs (y :: ys) (y :: xys)
+
+||| A record for storing the result of splitting a list `xys` according to some
+||| property `p`.
+||| The `prfs` and `contras` are related to the original list (`xys`) via
+||| `Interleaving`.
+|||
+||| @ xys the list which has been split
+||| @ p   the property used for the split
+public export
+record Split {a : Type} (p : a -> Type) (xys : List a) where
+  constructor MkSplit
+  {xs, ys : List a}
+  {auto interleaving : Interleaving xs ys xys}
+  ||| A proof that all elements in `xs` satisfies the property used for the
+  ||| split.
+  prfs    : All p xs
+  ||| A proof that all elements in `ys` do not satisfy the property used for the
+  ||| split.
+  contras : All (Not . p) ys
```

Comment:
oof yeah you'll probably want to use long names for these to avoid getting warnings about shadowing
for all of the uses of xs and ys in type signatures

Reply:
I took the names from the original Discord version for `xs` and `ys`: `ayes` and `naws` respectively.
Not great, not terrible? ^^